### PR TITLE
Fix REST query parameter bug

### PR DESCRIPTION
**Pull request by alouie-sfdc**
This is a fix for https://github.com/ryanbrainard/forceworkbench/issues/581.

When passing an "url" parameter to the REST Explorer, that parameter needs to have its ampersands URL-encoded so that they aren’t treated as separate query parameters in the top-level request.

### CODE REVIEW

**Comment by ryanbrainard** [comment]
Thanks for the PR. I was just trying this out on the [review app](https://forceworkbench-staging-pr-663.herokuapp.com), but doesn't seem to be totally fixed. 

For example, I go to:

    https://forceworkbench-staging-pr-663.herokuapp.com/restExplorer.php?url=/services/data/v34.0/chatter/feeds/news/00560000000wX0aAAE/feed-elements?pageSize=2

and then click on:

    updatesUrl: /services/data/v34.0/chatter/feeds/news/00560000000wX0aAAE/feed-elements?pageSize=2&updatedSince=2%3A1442468950000

The browser's URL correctly changes to:

    https://forceworkbench-staging-pr-663.herokuapp.com/restExplorer.php?url=/services/data/v34.0/chatter/feeds/news/00560000000wX0aAAE/feed-elements?pageSize=2&updatedSince=2%3A1442468950000&autoExec=1

But the path in the Workbench UI only changes to (without `updatedSince=2%3A1442468950000`):

    /services/data/v34.0/chatter/feeds/news/00560000000wX0aAAE/feed-elements?pageSize=2

**Comment by alouie-sfdc** [comment]
Hi Ryan, this is very strange. I was able to repro the problem on the review app (https://forceworkbench-staging-pr-663.herokuapp.com), but I couldn't repro it on my own app (https://alouie-forceworkbench.herokuapp.com/). I'm not sure what the discrepancy could be.

I'm going to make another commit for another minor issue shortly. Let's see if things get any better after that.

**Comment by alouie-sfdc** [comment]
@ryanbrainard, I've tacked on another minor enhancement to allow making POST, PATCH, and PUT requests without supplying a request body (it only applies to JSON input). My colleagues and I have hit this a number of times -- we want to make a valid request that uses query parameters, like this:

    POST /services/data/v34.0/chatter/feed-elements?subjectId=me&text=hi

... but we get an error because there's no request body.

The workaround is to type `null` into the body to indicate that the JSON input is null. My fix is to set `null` as the value if nothing's specified, and the Content-Type is application/json.

**Comment by nosoftwarepaul** [comment]
@ryanbrainard Any chance this PR will get merged in?

**Comment by nosoftwarepaul** [comment]
@dmgerow  would you be able to merge this pull request?

File: workbench/controllers/RestExplorerController.php [hunk 1]
```diff
@@ -70,7 +70,13 @@ public function preExecute() {
         $this->url = str_replace(' ', '+', trim($this->url));
 
         if (in_array($this->requestMethod, RestApiClient::getMethodsWithBodies()) && trim($this->requestBody) == "") {
-            throw new WorkbenchHandledException("Must include a Request Body.");
+            if (stripos($this->requestHeaders, "Content-Type: application/json") !== false) {
+                // If nothing's specified in the JSON body, set it to null.
+                $this->requestBody = "null";
```
Comment (ryanbrainard) [inline workbench/controllers/RestExplorerController.php]: How about setting this to `{}` instead? That's the usual (but not standard) way to indicate nothingness in JSON other than true `null`, which I'm guess won't work here.

File: workbench/controllers/RestResponseInstrumenter.php [hunk 2]
```diff
@@ -10,11 +10,16 @@ public function __construct($baseUrl) {
...
         $this->insturmentations[] = new RegexInsturmentation("https?://.*(/id/00D.*/005.*)",
-                                                 		     "<a class=\'RestLinkable\' href=" . $baseUrl . "?url=$1&autoExec=1>$1</a>");
+                                                             "<a class=\'RestLinkable\' href=" . $baseUrl . "?url=$1&autoExec=1>$1</a>");
 
         // add link url to any rest url
-        $this->insturmentations[] = new RegexInsturmentation("(" . $restBasePattern . ".*)",
-                                                 		     "<a class=\'RestLinkable\' href=" . $baseUrl . "?url=$1&autoExec=0>$1</a>");
+        $this->insturmentations[] = new RegexUrlInsturmentation("(" . $restBasePattern . ".*)",
+                                                                function($matches) {
+                                                                    // This is a fix for https://github.com/ryanbrainard/forceworkbench/issues/581
```
Comment (ryanbrainard) [inline workbench/controllers/RestResponseInstrumenter.php]: no need for this comment

File: workbench/controllers/RestResponseInstrumenter.php [hunk 3]
```diff
@@ -90,4 +95,21 @@ function instrument($s) {
         return preg_replace('@' . $this->pattern . '@', $this->replacement, $s);
     }
 }
-?>
\ No newline at end of file
+
+class RegexUrlInsturmentation implements Insturmentation {
```
Comment (ryanbrainard) [inline workbench/controllers/RestResponseInstrumenter.php]: why is this special for urls? seems like it's special that it takes a callback. perhaps rename the class or the param?

Comment (ryanbrainard) [inline workbench/controllers/RestResponseInstrumenter.php]: if this is for URLs, is should all the `urlencoded` business above happen inside here

File: workbench/controllers/RestResponseInstrumenter.php [hunk 4]
```diff
@@ -10,11 +10,16 @@ public function __construct($baseUrl) {
...
-                                                 		     "<a class=\'RestLinkable\' href=" . $baseUrl . "?url=$1&autoExec=1>$1</a>");
+                                                             "<a class=\'RestLinkable\' href=" . $baseUrl . "?url=$1&autoExec=1>$1</a>");
 
         // add link url to any rest url
-        $this->insturmentations[] = new RegexInsturmentation("(" . $restBasePattern . ".*)",
-                                                 		     "<a class=\'RestLinkable\' href=" . $baseUrl . "?url=$1&autoExec=0>$1</a>");
+        $this->insturmentations[] = new RegexUrlInsturmentation("(" . $restBasePattern . ".*)",
+                                                                function($matches) {
+                                                                    // This is a fix for https://github.com/ryanbrainard/forceworkbench/issues/581
+                                                                    $urlencoded = str_replace('&amp;', '%26', $matches[1]);
```
Comment (ryanbrainard) [inline workbench/controllers/RestResponseInstrumenter.php]: any way a user could inject malicious something in here from the URL?

**Comment by ntotten** [comment]
Closing as there hasn't been activity in more than a year.

**Pull request closed without merging**